plugin/src/main/resources/org/jenkinsci/plugins/workflow/cps/config/CPSConfiguration/help-showSandbox.html

```diff
@@ -0,0 +1,5 @@
+<div>
+    <p>Controls whether the "Use Groovy Sandbox" is shown in pipeline jobs configuration to users without Overall/Administer permission.</p>
+    <p>Note that this is not a security configuration, as it does not prevent users to configure and run
+        pipelines in some other ways without using the sandbox.</p>
```

Comment:
> Throw something from the @DataBoundConstructor if !sandbox.

It should also check if the current user is an admin (as they are allowed to disable the sandbox), but jobs can be created by other means which do not necesasrily have a user context (`job-dsl` for example, or CasC for items when using CloudBees CI). So this would not be a complete solution either. So given that there is not a complete solution, I decided to fix the minimum required to solve the UX issue that happens when using the UI.

Reply:
> jobs can be created by other means which do not necessarily have a user context (`job-dsl` for example)

In this case the person configuring the DSL job (or approving changes to its source code) is effectively a Jenkins admin whom we trust to decide whether to allow the sandbox to be disabled. Checking `hasPermission` would work uniformly, since `SYSTEM` would pass.

Reply:
I've been trying to block item creation/update by failing on the `@DataBoundConstructor` of `CpsFlowDefinition`. However UX is horrible, getting _Angry Jenkins_ (no matter what exception type I use - tried with `RuntimeException` and `Failure`, same result).

An alternative solution would be to unconditionally set `sandbox = true` when the right conditions are met (`CPSConfiguration.get().isHideSandbox() && !sandbox && !Jenkins.get().hasPermission(Jenkins.ADMINISTER)`). But then the sandbox would be _silently_ enabled when non-admin users save a job which had the sandbox disabled previously.

 Given all that, I think we should not open that can of worms and just hide the checkbox (current state of the PR).

Reply:
(Discussed out of band: core might have seen a regression in handling of `Failure` as of https://github.com/jenkinsci/jenkins/pull/4389. https://github.com/jenkinsci/jenkins/blob/cc97ad8f330aff43f412a95235a5386194803b14/core/src/main/java/hudson/model/Failure.java#L36-L41 is supposed to render https://github.com/jenkinsci/jenkins/blob/cc97ad8f330aff43f412a95235a5386194803b14/core/src/main/resources/hudson/model/AbstractModelObject/error.jelly#L33-L41; that is the whole purpose of this class.)

Reply:
Reverting that particular change generates a lot of non-trivial conflicts. So I've _manually_ reverted the main chunks in there and the behaviour is still not the expected. The _oops_ page is rendered (including the stacktrace). `Failure` is not behaving different than any other unhandled exception. I'm not sure how this would have ever worked, as it rendered the `oops.jelly` since [it's creation](https://github.com/jenkinsci/jenkins/commit/6badbb3227701f93aeebf208cc47f8b1540617da).

Reply:
`Failure` still works in general, including for non-admin users. Checked by trying to copy a nonexistent job as a user lacking `ADMINISTER`; rendered the polite error page with message as expected.
![failure](https://github.com/user-attachments/assets/1ab120b8-c95e-40c4-82fa-88536c23955a)


```
java.lang.Exception: Stack trace
	at java.base/java.lang.Thread.dumpStack(Thread.java:1389)
	at hudson.model.Failure.<init>(Failure.java:56)
	at hudson.model.Failure.<init>(Failure.java:50)
	at hudson.model.ItemGroupMixIn.createTopLevelItem(ItemGroupMixIn.java:179)
	at jenkins.model.Jenkins.doCreateItem(Jenkins.java:4281)
	at jenkins.model.Jenkins.doCreateItem(Jenkins.java:344)
	at hudson.model.AllView.doCreateItem(AllView.java:98)
	at …
```

rendered in

```
java.lang.Exception: Stack trace
	at java.base/java.lang.Thread.dumpStack(Thread.java:1389)
	at hudson.model.Failure.generateResponse(Failure.java:68)
	at org.kohsuke.stapler.HttpResponseRenderer$Default.handleHttpResponse(HttpResponseRenderer.java:129)
	at org.kohsuke.stapler.HttpResponseRenderer$Default.generateResponse(HttpResponseRenderer.java:71)
	at org.kohsuke.stapler.Function.renderResponse(Function.java:162)
	at org.kohsuke.stapler.Function.bindAndInvokeAndServeResponse(Function.java:152)
	at org.kohsuke.stapler.MetaClass$11.doDispatch(MetaClass.java:644)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:61)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:804)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:942)
	at org.kohsuke.stapler.MetaClass$4.doDispatch(MetaClass.java:327)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:61)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:804)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:942)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:725)
	at org.kohsuke.stapler.Stapler.service(Stapler.java:253)
	at …
```

What was the user gesture you were testing, and what was its stack trace? Perhaps some intermediate method was catching `Failure` and rethrowing something else, or perhaps it was from an endpoint which somehow bypasses `HttpResponseRenderer`.

Reply:
```diff
diff --git a/plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition.java b/plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition.java
index aeae1865..153b2352 100644
--- a/plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition.java
+++ b/plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition.java
@@ -91,6 +91,7 @@ public class CpsFlowDefinition extends FlowDefinition {
         this.script = sandbox ? script : ScriptApproval.get().configuring(script, GroovyLanguage.get(),
                 ApprovalContext.create().withCurrentUser().withItemAsKey(req != null ? req.findAncestorObject(Item.class) : null), req == null);
         this.sandbox = sandbox;
+        throw new Failure("Cannot save!!");
     }
 
     private Object readResolve() {
```
And try to save a pipeline.

Reply:
Ah, because it is wrapped in `Descriptor.newInstance` as you can see in the stack trace. I tried `Descriptor.FormException`, which is supposed to be for this purpose. (In fact we could even revert the code to hide the checkbox—this is arguably better UX.) Unfortunately right now it is getting wrapped by `RequestImpl.invokeConstructor` then `RequestImpl.TypePair.convertJSON`. Seems like a Stapler bug :eyes:

Reply:
Picked https://github.com/jenkinsci/stapler/pull/571 + https://github.com/jenkinsci/jenkins/pull/9495 up. It's working.

However, I just realized that we can't block the `create-job` CLI as it's not going through the data-bound constructor.

Reply:
Yes, both of those would go through `readResolve` as called by XStream. At least the CLI will treat certain exception types like `IllegalArgumentException` specially for purposes of error handling, if it matters.